src/policy/policy_deny_fs.cc

```diff
@@ -0,0 +1,137 @@
+    Permission perm = Permission::kPermissionsRoot;
+    for (std::string& opt : SplitString(name, ':')) {
+      if (perm == Permission::kPermissionsRoot) {
+        if (opt == "fs") {
+          deny_all_in_ = true;
+          deny_all_out_ = true;
+          return Just(true);
+        }
+        if (opt == "in") {
+          perm = Permission::kFileSystemIn;
+          deny_all_in_ = true;
+        } else if (opt == "out") {
+          perm = Permission::kFileSystemOut;
+          deny_all_out_ = true;
+        } else {
+          return Just(false);
+        }
+      } else {
+        RestrictAccess(perm, opt);
+      }
+    }
+  }
+
+  return Just(true);
+}
+
+bool PolicyDenyFs::Deny(Permission perm, std::vector<std::string> params) {
+  if (perm == Permission::kFileSystem) {
+    deny_all_in_ = true;
+    deny_all_out_ = true;
+    return true;
+  }
+
+  bool deny_all = params.size() == 0;
+  if (perm == Permission::kFileSystemIn) {
+    if (deny_all) deny_all_in_ = true;
+    // when deny_all_in is already true policy.deny should be idempotent
+    if (deny_all_in_) return true;
+
+    RestrictAccess(perm, params);
+    return true;
+  }
+
+  if (perm == Permission::kFileSystemOut) {
+    if (deny_all) deny_all_out_ = true;
+    // when deny_all_out is already true policy.deny should be idempotent
+    if (deny_all_out_) return true;
+
+    RestrictAccess(perm, params);
+    return true;
+  }
+
+  return false;
+}
+
+void PolicyDenyFs::RestrictAccess(Permission perm, const std::string& res) {
+  char resolvedPath[PATH_MAX];
+  // check the result
+  realpath(res.c_str(), resolvedPath);
+
+  std::filesystem::path path(resolvedPath);
+  bool isDir = std::filesystem::is_directory(path);
+  // when there are parameters deny_params_ is automatically
+  // set to false
+  if (perm == Permission::kFileSystemIn) {
+    deny_all_in_ = false;
+    deny_in_params_.push_back(std::make_pair(resolvedPath, isDir));
+  } else if (perm == Permission::kFileSystemOut) {
+    deny_all_out_ = false;
+    deny_out_params_.push_back(std::make_pair(resolvedPath, isDir));
+  }
+}
+
+void PolicyDenyFs::RestrictAccess(Permission perm,
+                                  std::vector<std::string> params) {
+  for (auto& param : params) {
+    RestrictAccess(perm, param);
+  }
+}
+
+bool PolicyDenyFs::is_granted(Permission perm, const std::string& param = "") {
+  switch (perm) {
+    case Permission::kFileSystem:
+      return !(deny_all_in_ && deny_all_out_);
+    case Permission::kFileSystemIn:
+      return !deny_all_in_ &&
+        (param.empty() || PolicyDenyFs::is_granted(deny_in_params_, param));
+    case Permission::kFileSystemOut:
+      return !deny_all_out_ &&
+        (param.empty() || PolicyDenyFs::is_granted(deny_out_params_, param));
+    default:
+      return false;
+  }
+}
+
+bool PolicyDenyFs::is_granted(DenyFsParams params, const std::string& opt) {
+  char resolvedPath[PATH_MAX];
+  realpath(opt.c_str(), resolvedPath);
```

Comment:
It would be good to discuss why this is using realpath and what security guarantees this gives that wouldn’t be there without. There is no single canonical path for files on POSIX systems, for example, not even via `realpath()`, or potentially no meaningful path at all. `realpath()` also might not return something that is intended to be interpreted as a file system path; for example:

```
$ realpath /proc/7495/fd/71
/dev/shm/.org.chromium.Chromium.GQUsqv (deleted)
```

the `(deleted)` string is part of the literal realpath() result. These things are fine for regular fs access, but for a security-relevant feature, this needs to be well thought through.

(As I mentioned to you, all of these are reasons why https://github.com/nodejs/node/pull/22112 didn’t attempt to tackle path-based file system access restrictions: it’s ******* hard to get right.)

Reply:
I've decided to use `realpath` mostly for consistency and it would make the radix-tree algorithm perform better. Honestly, the behavior you described seems expected and the documentation should elaborate/clarify those situations. Regarding the security concerns, I'll definitely need to evaluate them with the @nodejs/security.

However, I'll change the `realpath` usage to a libuv call same as `path.resolve`. So, I'd say the security concern wouldn't be much more than an existing behavior in `path.resolve`.

Reply:
> a libuv call same as `path.resolve`

Not quite sure I understand this – `path.resolve()` doesn’t do libuv calls, other than maybe one to get the current working directory.

Reply:
Sorry, I meant: `uv_fs_realpath`

Reply:
Alright – that seems like a good idea. But still, keep this in mind – using an fs call here means that you need to be careful about how to propagate errors from it, so that an application cannot test e.g. whether a file exists on a path that it should not have access to by letting this call here do the check for it.

Reply:
Ok, I gave it a shot and I found a barrier:

`uv_fs_realpath` is equivalent to `realpath(3)` and as described by the docs it returns:

> `ENOENT` when the named file does not exist.

And the `is_granted` method can be called to write into a file that doesn't exist yet (so it will create), then this call will fail (causing `req.ptr` == nullptr). Therefore, I guess we shouldn't use it in this path at least.

Also, it's important to mention that in the libuv docs there's a mention of a few edge cases when using `uv_fs_realpath`:

> Warning This function has certain platform-specific caveats that were discovered when used in Node. Ref: https://docs.libuv.org/en/latest/fs.html?highlight=uv_fs_realpath#c.uv_fs_realpath

---

I also believe using `realpath(1)` (current implementation) is not a good approach, any idea?

Reply:
I’d not do any path translations here for now, including no `realpath()`. But either way, this needs to be well-documented and thought through.

Reply:
Hi Anna! I'm doing some experiments and I'm trying to avoid the `path.resolve()` on the JS side as much as I can. The reason is that it doesn't seem the _best_ approach to me. I mean, every `fs.*` call would need to swap between native code even when there is no `policy.deny` arguments. Of course, I can use a variable when no arguments are set to that specific permission, but IMHO it would add unnecessary complexity.

I'm looking to other approaches, and how do you feel about implementing the [`path.resolve`](https://github.com/nodejs/node/blob/main/lib/path.js#L1091)  on the C++ side as well? (including posix and win implementation)

Reply:
> I mean, every `fs.*` call would need to swap between native code even when there is no `policy.deny` arguments.

Wouldn’t this still be the case if you implement `path.resolve()` in C++? I mean, sure, you can do that, but it seems equally complex, with the addition of having to port a method to C++ and keep it in sync.

Reply:
> Wouldn’t this still be the case if you implement path.resolve() in C++?

The difference is that it will be called just when is needed, inside `policy_deny_fs.cc`. I wouldn't need to check if it should be called in every `fs.*` call

Reply:
I’m not sure I understand – the minimum amount of change needed per fs function here is one extra function call to potentially resolve the file path(s) to an absolute one. The complexity of deciding whether to apply the transformation or not can be encapsulated inside that function call, but it won’t go away, and it will be the same regardless of where it happens.

Advantages of doing it in C++ would be that you don’t need to worry about string vs buffer paths and that it might potentially simplify the check in the permissions code that the path is an absolute one, advantages of doing it in JS would be that there is an existing implementation already available.

Reply:
FWIW We've discussed it in private and sounds reasonable to give C++ `path.resolve` a shot. In case that prove to be not worth it (performance-wise), I can use the JS approach:

```js
function possiblyTransformPath(path) {
  if (pathModule.isAbsolute(path)) return path;
  if (process.policy.hasAnyConfig('fs.in') || process.policy.hasAnyConfig('fs.out')) {
    path = pathModule.resolve(path);
  }
  return path;
}
```